### PR TITLE
Optimize UrlEncode in Utils

### PR DESCRIPTION
**Pull request by danielmarbach**
## Description

Optimizes UrlEncode for better memory efficiency

## Motivation and Context

```

BenchmarkDotNet v0.13.12, macOS Sonoma 14.4.1 (23E224) [Darwin 23.4.0]
Apple M3 Max, 1 CPU, 14 logical and 14 physical cores
.NET SDK 8.0.202
  [Host]     : .NET 8.0.3 (8.0.324.11423), Arm64 RyuJIT AdvSIMD
  DefaultJob : .NET 8.0.3 (8.0.324.11423), Arm64 RyuJIT AdvSIMD


```

image

https://github.com/danielmarbach/aws-sdk-net-benchmarks/blob/main/UrlEncode.cs

## Testing

Added tests to verify the code still does the same as before. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: sdk/src/Core/AWSSDK.Core.NetStandard.csproj
Comment (danielmarbach) [inline sdk/src/Core/AWSSDK.Core.NetStandard.csproj:27]: Like discussed before this is required for some of the shim APIs.

File: sdk/src/Core/Amazon.Runtime/Internal/Util/_netstandard/EncodingExtensions.cs
Comment (danielmarbach) [inline sdk/src/Core/Amazon.Runtime/Internal/Util/_netstandard/EncodingExtensions.cs:26]: Let me know if you prefer another place to put it

File: sdk/src/Core/Amazon.Runtime/Internal/_bcl+netstandard/SkipLocalsInitAttribute.cs
Comment (danielmarbach) [inline sdk/src/Core/Amazon.Runtime/Internal/_bcl+netstandard/SkipLocalsInitAttribute.cs:6]: Let me know if you prefer another place to put it

File: sdk/src/Core/Amazon.Util/AWSSDKUtils.cs
Comment (danielmarbach) [inline sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:78]: I did check the entire SDK and couldn't find another reference so I got rid of it. Is it possible to entirely get rid of the different RFC implementations? If that is possible we could potentially even get rid of the custom code and replace it with the System.Net implementation? 

https://learn.microsoft.com/en-us/dotnet/api/system.net.webutility.urlencode?view=net-8.0

File: sdk/test/NetStandard/UnitTests/AWSSDK.UnitTests.Custom.NetStandard.csproj
Comment (danielmarbach) [inline sdk/test/NetStandard/UnitTests/AWSSDK.UnitTests.Custom.NetStandard.csproj:9]: Netstandard is not a valid target for unit tests

File: sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs
Comment (danielmarbach) [inline sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:21]: Should I add more data from the valid path chars?

File: sdk/src/Core/Amazon.Util/AWSSDKUtils.cs
Comment (danielmarbach) [inline sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:1116]: Given the encoding schemes are only used here the string concat could also go by already have the concatenated string available as costs and returning them by TryGetRFCEncodingSchemes

File: sdk/src/Core/Amazon.Util/AWSSDKUtils.cs
Comment (danielmarbach) [inline sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:1122]: I left the 2 * dataLength in here since the previous method made similar assumptions.I assume this was chosen because it is unrealistic to only have data that will have to be encoded all the way like a string with only spaces or for example a string like "!!!!"

Comment (danielmarbach) [reply sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:1122]: image

Comment (danielmarbach) [reply sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:1122]: Never mind we should be good with 2 * data length

**Comment by danielmarbach** [comment]
@normj @dscpinheiro I know this is still draft but would you mind having a look at it at some point? I have added a bunch of inline comments that might influence this PR.

**Comment by danielmarbach** [comment]
Ping @normj

File: sdk/src/Core/AWSSDK.Core.NetStandard.csproj
Comment (danielmarbach) [reply sdk/src/Core/AWSSDK.Core.NetStandard.csproj:27]: I don't understand this argument
The net runtime allows unsafe and runs pretty much on any platform. Azure SDK allows unsafe and runs on a variety of platforms too. Rabbitmq client which I contribute too is used in all sorts of environments and allows unsafe too. With the choice of your TFMs you will have to backport some amount of helpers that sometimes require "unsafe" modifier in order to balance getting perf and also reducing the complexity of the code base by reducing the number of redundant code paths that have to be implemented across the board.

Unsafe is unsafe in the sense that you need to know what you are doing and yes ultimately there should be as little as possible unsafe code but that should not mean there can't be any because of a fear of making a mistake.

Fyi the code here was ported from NET originally and brought into the Azure SDK and has been battle tested against a plethora of targets over years

File: sdk/src/Core/Amazon.Util/AWSSDKUtils.cs
Comment (normj) [inline sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:1126]: The wording of this comment is confusing. Maybe something like

```
// Instead of stack allocating or renting two buffers we use one buffer with at least twice the capacity of the
// max encoding length. Then store the character data as bytes in the second half reserving the first half of the buffer 
// for the encoded representation.
```

File: sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs
Comment (normj) [inline sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:10]: Can we add some tests with multi byte characters?

File: sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs
Comment (normj) [reply sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:21]: You have proven that the characters from the valid path are being taken into account. I don't think this test needs to test each value that is in the collection.

File: sdk/src/Core/Amazon.Runtime/Internal/Util/_netstandard/EncodingExtensions.cs
Comment (normj) [reply sdk/src/Core/Amazon.Runtime/Internal/Util/_netstandard/EncodingExtensions.cs:26]: You can add these extensions method to this .NET Standard extensions file. https://github.com/aws/aws-sdk-net/blob/main/sdk/src/Core/Amazon.Util/_netstandard/Extensions.cs

**Review by normj — CHANGES_REQUESTED** [review]

File: sdk/src/Core/Amazon.Runtime/Internal/Util/_netstandard/EncodingExtensions.cs
Comment (danielmarbach) [reply sdk/src/Core/Amazon.Runtime/Internal/Util/_netstandard/EncodingExtensions.cs:26]: Done

File: sdk/src/Core/Amazon.Runtime/Internal/_bcl+netstandard/SkipLocalsInitAttribute.cs
Comment (danielmarbach) [reply sdk/src/Core/Amazon.Runtime/Internal/_bcl+netstandard/SkipLocalsInitAttribute.cs:6]: Done

File: sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs
Comment (danielmarbach) [reply sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:21]: I wanted to play extra safe TBH

File: sdk/src/Core/Amazon.Util/AWSSDKUtils.cs
Comment (danielmarbach) [reply sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:1126]: Haha. My bad English shines through 👯

Comment (danielmarbach) [reply sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:1126]: Done

File: sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs
Comment (danielmarbach) [reply sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:10]: Done

**Comment by danielmarbach** [comment]
I raised https://github.com/aws/aws-sdk-net/issues/3326 because the constants are a bit messy and need an overhaul. I did my best to get things into compilable state for now.

**Comment by danielmarbach** [comment]
Ok I'll do a rebase

**Comment by danielmarbach** [comment]
@normj rebased

**Review by normj — APPROVED** [review]
Looks good. I'll get somebody else on the team to do a second pass.

**Comment by danielmarbach** [comment]
Rebased to solve the conflict

**Comment by danielmarbach** [comment]
Would be nice if this could be merged since I have a bunch of other PRs that introduce more tests to the test file that probably require a rebase then ;)

**Comment by boblodgett** [comment]
> Would be nice if this could be merged since I have a bunch of other PRs that introduce more tests to the test file that probably require a rebase then ;)

Taking a look

**Comment by boblodgett** [comment]
I don't visually see issues in the PR. However, this didn't build in our build system because a change is required to support the /unsafe switch. I will make the change so we can ensure this PR will build and test correctly before I approve the PR.

File: sdk/src/Core/AWSSDK.Core.NetStandard.csproj
Comment (normj) [inline sdk/src/Core/AWSSDK.Core.NetStandard.csproj:27]: @danielmarbach can you add this same line to the `AWSSDK.Core.NetFramework.csproj` file? The internal builds are failing for the .NET Framework project because the PR doesn't allow unsafe code in .NET Framework.

**Review by normj — CHANGES_REQUESTED** [review]

File: sdk/src/Core/AWSSDK.Core.NetStandard.csproj
Comment (danielmarbach) [reply sdk/src/Core/AWSSDK.Core.NetStandard.csproj:27]: Sure. FYI all my PRs have push right

File: sdk/test/NetStandard/UnitTests/AWSSDK.UnitTests.Custom.NetStandard.csproj
Comment (boblodgett) [reply sdk/test/NetStandard/UnitTests/AWSSDK.UnitTests.Custom.NetStandard.csproj:9]: Ran the build system again and got the following error:

```
C:\Program Files\dotnet\sdk\8.0.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1005: Assets file 'C:\codebuild\tmp\output\src3842370628\src\aws-sdk-net\sdk\test\NetStandard\UnitTests\obj\AWSSDK.UnitTests.Custom.NetStandard\project.assets.json' doesn't have a target for 'netstandard2.0'. Ensure that restore has run and that you have included 'netstandard2.0' in the TargetFrameworks for your project.
```

I believe it is likely because of this change which needs to be reverted. Although you are correct that it really isn't a correct target to run for unit tests it is part of the solution. So when the solution is built with:

```
dotnet build -c Release --no-incremental $buildTarget -f netstandard2.0 /p:AWSKeyFile=$CustomSnkLocation /p:SignAssembly=true
```

The `-f netstandard2.0` switch which we used for parallel target building forces that it finds every project with that target in the solution. Since we have a multitargeted NetStandard.csproj and a single sln for Netstandard that doesn't have netstandard2.0 broken out it is required here.

**Review by boblodgett — CHANGES_REQUESTED** [review]
https://github.com/aws/aws-sdk-net/pull/3307#discussion_r1664600695

File: sdk/test/NetStandard/UnitTests/AWSSDK.UnitTests.Custom.NetStandard.csproj
Comment (danielmarbach) [reply sdk/test/NetStandard/UnitTests/AWSSDK.UnitTests.Custom.NetStandard.csproj:9]: Ok I'm going to revert this for now to not further derail things but I have questions 😂

Comment (danielmarbach) [inline sdk/test/NetStandard/UnitTests/AWSSDK.UnitTests.Custom.NetStandard.csproj:9]:
```suggestion
        <TargetFrameworks>netstandard2.0;netcoreapp3.1;net8.0</TargetFrameworks>
```

Comment (danielmarbach) [reply sdk/test/NetStandard/UnitTests/AWSSDK.UnitTests.Custom.NetStandard.csproj]: Done

**Review by normj — APPROVED** [review]
Approve with assumption @boblodgett gets a successful build through the internal system.

**Comment by dscpinheiro** [comment]
@danielmarbach I approved and merged #3363 and #3365, but we saw integration test failures when running this PR on our build system (they were all related to this CloudFormation section: https://github.com/aws/aws-sdk-net/blob/3.7.844.0/sdk/test/NetStandard/IntegrationTests/IntegrationTests/CloudFormation.cs#L20).

Essentially that value is passed to `UrlEncode` (via `GetParametersAsString`), but `AWSSDKUtils.UrlEncode(TEMPLATE_TEXT, false)` in your branch is not returning the same value as `v4-development`.

**Comment by danielmarbach** [comment]
@dscpinheiro Ok I'll check

**Comment by danielmarbach** [comment]
@dscpinheiro I'm approaching my leave period so it might take a bit longer for me to respond. Apologies in advance

**Comment by danielmarbach** [comment]
@dscpinheiro I fixed the problem. It was between the chair and the keyboard 👯

File: sdk/src/Core/Amazon.Util/AWSSDKUtils.cs
Comment (danielmarbach) [inline sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:1057]: Technically we could potentially remove bound checks here by doing something like 

```
Unsafe.Add(ref MemoryMarshal.GetReference(dataBuffer), index++) = symbol;
```

but I decided to not go down that path since it increases the complexity of the code

**Comment by danielmarbach** [comment]
Give it another run through the benchmark

image

Still very good allocation reduction and the throughput hit is still in very acceptable range

**Comment by boblodgett** [comment]
All the new changes are being run through the backend build system now.

**Comment by boblodgett** [comment]
One test failure for netcoreapp3.1 target running tests.

```
error : Test failure in XUnitTestRunner: UnitTests.NetStandard.Core.AWSSDKUtilsTests.UrlEncodeWithoutPath(input: "{\r\n\t\"AWSTemplateFormatVersion\" : \"2010-09-09"..., expected: "%7B%0A%09%22AWSTemplateFormatVersion%22%20%3A%20%2"...)
```

**Comment by danielmarbach** [comment]
I ran against net8 and as far as I recall net4.x. I'm not installing a runtime that is out of support for almost two years and hasn't received security updates since them. Sorry. 

So if you can provide me the entire test output with some information that makes it possible for me to guess what's going on I'm happy to give it another look.

Otherwise this PR has also edit rights. Feel free to push updates.

**Comment by boblodgett** [comment]
> I ran against net8 and as far as I recall net4.x. I'm not installing a runtime that is out of support for almost two years and hasn't received security updates since them. Sorry.
> 
> So if you can provide me the entire test output with some information that makes it possible for me to guess what's going on I'm happy to give it another look.
> 
> Otherwise this PR has also edit rights. Feel free to push updates.


Thank you for working with us on this PR while we shake out the test issues. I ran the test in Visual Studio locally and it also fails in .NET 8.

The project containing the failed test is: `.\aws-sdk-net\sdk\test\NetStandard\UnitTests\AWSSDK.UnitTests.Custom.NetStandard.csproj`

The file containing the failed test within that project is: `.\aws-sdk-net\sdk\test\NetStandard\UnitTests\Core\AWSSDKUtilsTests.cs`

The test is: `public void UrlEncodeWithoutPath(string input, string expected)` which has three [InlineData] test cases. The case that is failing specifically is the third one which uses the `TemplateText`.

Here is a link to the test case: https://github.com/danielmarbach/aws-sdk-net/blob/f5fe55e95b98d07bb7cd2c2a8e8f4451f0b1b819/sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs#L70

TemplateText can be found here: https://github.com/danielmarbach/aws-sdk-net/blob/f5fe55e95b98d07bb7cd2c2a8e8f4451f0b1b819/sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs#L78

All other tests in this project pass.

**Comment by boblodgett** [comment]
From the look of it `\r` (0x0D) characters are being injected and encoded when the test case hits a `\n` (0x0A):

image

File: sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs
Comment (boblodgett) [inline sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:70]: This is the test case that is failing which I didn't remember was a new test case until reviewing the changes again. This likely passes on a Linux based machine but will fail on Windows because of how some of the methods automatically translate `\n` to `\r\n` in windows.

**Review by boblodgett — CHANGES_REQUESTED** [review]

File: sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs
Comment (danielmarbach) [reply sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:70]: Ah yeah the input needs to be normalized. I can only do that after my vacation. Feel free to push it if you have time.

Comment (danielmarbach) [reply sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:70]: Does it happen on checkout on the build agent?

**Comment by danielmarbach** [comment]
Honestly looking at this again on my phone I'd argue the only thing that would have been needed to cover the bug in https://github.com/aws/aws-sdk-net/pull/3307/commits/618ef3e80e8748c4f40162fe61bc95ba446ed521 would have been a long enough string. Fighting line endings and going back and forth on this PR for trying to fit the template text into here seems a bit pointless and unnecessarily holding this up. How about I remove this test case for now?

File: sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs
Comment (danielmarbach) [inline sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:70]:
```suggestion
```

Comment (boblodgett) [reply sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:70]: Yes, there are going to be problems with the line endings depending on where the test is run. I would have to dig into seeing if the build system or where it is even happening on my Windows system but it will easily get switched to `\r\n` when a developer touches this file on Windows. Thus this test case is too fragile and should be removed.

Comment (boblodgett) [reply sdk/test/NetStandard/UnitTests/Core/AWSSDKUtilsTests.cs:70]: I will strip that test case and run it again to make sure nothing else it blocking in the meantime.

**Comment by danielmarbach** [comment]
@boblodgett based on the above comment I removed the template test for now. Having some tests in place for the method where there previously have been none is already a good start and the template was only a temporary thing anyway which helped me discover the buffer length problem. Ideally, there would be another test with a longer input but I cannot do this right now since I'm not around my laptop or workstation for a while

**Comment by danielmarbach** [comment]
And thanks for your support here and sorry for me being less helpful this time

**Review by boblodgett — APPROVED** [review]

**Pull request merged**
